src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests_ListPatterns.cs

```diff
@@ -7638,4 +7815,116 @@ public void Simple_Array_Slice_VerifyRangeMaths(string data, string pattern, str
+        comp.VerifyEmitDiagnostics(
+            // (2,34): error CS0021: Cannot apply indexing with [] to an expression of type 'INotCountable'
+            // _ = new C() is INotCountable and [var x, .. var y];
+            Diagnostic(ErrorCode.ERR_BadIndexLHS, "[var x, .. var y]").WithArguments("INotCountable").WithLocation(2, 34),
+            // (2,42): error CS0021: Cannot apply indexing with [] to an expression of type 'INotCountable'
+            // _ = new C() is INotCountable and [var x, .. var y];
+            Diagnostic(ErrorCode.ERR_BadIndexLHS, ".. var y").WithArguments("INotCountable").WithLocation(2, 42)
+            );
+    }
+
+    [Fact]
+    public void ListPattern_ExplicitInterfaceImplementation()
+    {
+        var source = @"
+if (new C() is Interface and [var x, .. var y])
+{
+    System.Console.Write((x, y));
+}
+
+interface Interface
+{
+    int Length { get; }
+    int this[System.Index i] { get; }
+    int this[System.Range r] { get; }
+}
+class C : Interface
+{
+    int Interface.Length { get => 2; }
+    int Interface.this[System.Index i] { get => 42; }
+    int Interface.this[System.Range r] { get => 43; }
+}
+";
+        var comp = CreateCompilation(new[] { source, TestSources.Index, TestSources.Range });
+        var verifier = CompileAndVerify(comp, expectedOutput: "(42, 43)");
+        verifier.VerifyDiagnostics();
+    }
+
+    [Fact]
+    public void ListPattern_Tuples()
+    {
+        var source = @"
+_ = (1, 2) is [var x, .. var y];
+
+System.Runtime.CompilerServices.ITuple ituple = default;
+_ = ituple is [var x2];
+_ = ituple is [..var y2];
+
+_ = ituple switch
+{
+    [1, 2] => 0,
+    (1, 2) => 0,
+    _ => 0,
+};
+";
+        var comp = CreateCompilation(new[] { source, TestSources.Index, TestSources.Range, TestSources.ITuple });
+        comp.VerifyEmitDiagnostics(
+            // (2,15): error CS0021: Cannot apply indexing with [] to an expression of type '(int, int)'
+            // _ = (1, 2) is [var x, .. var y];
+            Diagnostic(ErrorCode.ERR_BadIndexLHS, "[var x, .. var y]").WithArguments("(int, int)").WithLocation(2, 15),
+            // (2,23): error CS0021: Cannot apply indexing with [] to an expression of type '(int, int)'
+            // _ = (1, 2) is [var x, .. var y];
+            Diagnostic(ErrorCode.ERR_BadIndexLHS, ".. var y").WithArguments("(int, int)").WithLocation(2, 23),
+            // (6,16): error CS1503: Argument 1: cannot convert from 'System.Range' to 'int'
+            // _ = ituple is [..var y2];
+            Diagnostic(ErrorCode.ERR_BadArgType, "..var y2").WithArguments("1", "System.Range", "int").WithLocation(6, 16)
+            );
+    }
+
+    [Fact]
+    public void ListPattern_NullTestOnSlice()
+    {
+        var source = @"
+using System;
+Span<int> s = default;
+switch (s)
+{
+    case [..[1],2,3]:
+    case [1,2,3]: // error
+        break;
+}
+
+int[] a = default;
+switch (a)
+{
+    case [..[1],2,3]:
+    case [1,2,3]: // no error
```

Comment:
>case [1,2,3]: // no error [](http://example.com/codeflow?start=4&length=25)

Why is there a behavior difference?

Reply:
Because we inject a null test for reference types (`int[]` vs. `Span<int>`) that we get back from the slice operation. We're having an email thread with LDM on this.